File: scripts/generate_thumbnail.py
#!/usr/bin/env python3
"""
Generate a thumbnail PNG via the OpenAI Images API and write it directly to disk,
along with a small sidecar .md noting the prompt/model/timestamp used.

Usage:
    python scripts/generate_thumbnail.py --project <name> --prompt "<text>"
    python scripts/generate_thumbnail.py --project <name> --prompt-file <path> [--out <custom.png>]
"""
import argparse
import base64
import datetime
from pathlib import Path

import requests

from common import openai_config, project_dir, fail

IMAGE_TIMEOUT_SECONDS = 3 * 60
IMAGE_SIZE = "1536x1024"  # matches the old app's real API call; concepts are framed as 16:9 in prose only


def parse_args():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("--project", required=True, help="Project name under projects/")
    prompt_source = parser.add_mutually_exclusive_group(required=True)
    prompt_source.add_argument("--prompt", help="Inline image prompt text")
    prompt_source.add_argument("--prompt-file", help="Path to a file containing the prompt text")
    parser.add_argument("--out", default=None, help="Output PNG path (default: projects/<project>/03_thumbnail_image.png)")
    return parser.parse_args()


def extract_error(response) -> str:
    try:
        body = response.json()
        return body.get("error", {}).get("message", response.text[:300])
    except ValueError:
        return response.text[:300]


def main():
    args = parse_args()
    cfg = openai_config()
    proj_dir = project_dir(args.project)

    if args.prompt_file:
        prompt = Path(args.prompt_file).read_text(encoding="utf-8").strip()
    else:
        prompt = args.prompt.strip()
    if not prompt:
        fail("Prompt is empty.")

    out_path = Path(args.out) if args.out else proj_dir / "03_thumbnail_image.png"
    sidecar_path = out_path.with_suffix(".md")

    url = f"{cfg['base_url']}/v1/images/generations"
    # No response_format param: gpt-image-1 rejects it (HTTP 400 "Unknown parameter")
    # and always returns b64_json by default, unlike older dall-e models.
    try:
        response = requests.post(
            url,
            headers={"Authorization": f"Bearer {cfg['api_key']}"},
            json={
                "model": cfg["image_model"],
                "prompt": prompt,
                "size": IMAGE_SIZE,
            },
            timeout=IMAGE_TIMEOUT_SECONDS,
        )
    except requests.RequestException as exc:
        fail(f"OpenAI image request failed: {exc}")

    if response.status_code >= 400:
        fail(f"OpenAI image generation failed (HTTP {response.status_code}): {extract_error(response)}")

    body = response.json()
    b64 = (body.get("data") or [{}])[0].get("b64_json")
    if not b64:
        fail("OpenAI did not return image data")

    image_bytes = base64.b64decode(b64)
    out_path.parent.mkdir(parents=True, exist_ok=True)
    out_path.write_bytes(image_bytes)

    timestamp = datetime.datetime.now(datetime.timezone.utc).isoformat(timespec="seconds")
    sidecar_path.write_text(
        "# Thumbnail image\n\n"
        f"- Model: {cfg['image_model']}\n"
        f"- Generated: {timestamp}\n"
        f"- Output: {out_path.name}\n\n"
        "## Prompt used\n\n"
        f"{prompt}\n",
        encoding="utf-8",
    )

    print(f"Generated {len(image_bytes)} bytes -> {out_path} (sidecar: {sidecar_path.name})")


if __name__ == "__main__":
    main()

File: scripts/generate_thumbnail.js
#!/usr/bin/env node
// Generate a thumbnail PNG via the OpenAI Images API and write it directly to disk,
// along with a small sidecar .md noting the prompt/model/timestamp used.
//
// Usage:
//   node scripts/generate_thumbnail.js --project <name> --prompt "<text>"
//   node scripts/generate_thumbnail.js --project <name> --prompt-file <path> [--out <custom.png>]
import fs from 'node:fs';
import path from 'node:path';
import { parseArgs } from 'node:util';
import { openaiConfig, projectDir, fail } from './common.js';

const IMAGE_TIMEOUT_MS = 3 * 60 * 1000;
const IMAGE_SIZE = '1536x1024'; // gpt-image-1's real size; thumbnail concepts are framed as 16:9 in prose only

function parseCliArgs() {
  const { values } = parseArgs({
    options: {
      project: { type: 'string' },
      prompt: { type: 'string' },
      'prompt-file': { type: 'string' },
      out: { type: 'string' },
    },
  });

  if (!values.project) {
    fail('--project is required.');
  }
  if (!values.prompt === !values['prompt-file']) {
    fail('Exactly one of --prompt or --prompt-file is required.');
  }

  return {
    project: values.project,
    prompt: values.prompt ?? null,
    promptFile: values['prompt-file'] ?? null,
    out: values.out ?? null,
  };
}

async function extractError(response) {
  const raw = await response.text();
  try {
    const body = JSON.parse(raw);
    return body?.error?.message ?? raw.slice(0, 300);
  } catch {
    return raw.slice(0, 300);
  }
}

async function main() {
  const args = parseCliArgs();
  const cfg = openaiConfig();
  const projDir = projectDir(args.project);

  const rawPrompt = args.promptFile
    ? fs.readFileSync(args.promptFile, 'utf8')
    : args.prompt;
  const prompt = rawPrompt.trim();
  if (!prompt) {
    fail('Prompt is empty.');
  }

  const outPath = args.out ? path.resolve(args.out) : path.join(projDir, '03_thumbnail_image.png');
  const { dir: outDir, name: outName } = path.parse(outPath);
  const sidecarPath = path.join(outDir, `${outName}.md`);

  const url = `${cfg.baseUrl}/v1/images/generations`;
  // No response_format param: gpt-image-1 rejects it (HTTP 400 "Unknown parameter")
  // and always returns b64_json by default, unlike older dall-e models.
  let response;
  try {
    response = await fetch(url, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${cfg.apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ model: cfg.imageModel, prompt, size: IMAGE_SIZE }),
      signal: AbortSignal.timeout(IMAGE_TIMEOUT_MS),
    });
  } catch (err) {
    fail(`OpenAI image request failed: ${err.message}`);
  }

  if (response.status >= 400) {
    fail(`OpenAI image generation failed (HTTP ${response.status}): ${await extractError(response)}`);
  }

  const body = await response.json();
  const b64 = body?.data?.[0]?.b64_json;
  if (!b64) {
    fail('OpenAI did not return image data');
  }

  const imageBytes = Buffer.from(b64, 'base64');
  fs.mkdirSync(path.dirname(outPath), { recursive: true });
  fs.writeFileSync(outPath, imageBytes);

  const timestamp = new Date().toISOString().replace(/\.\d{3}Z$/, '+00:00');
  const sidecar = `# Thumbnail image\n\n- Model: ${cfg.imageModel}\n- Generated: ${timestamp}\n- Output: ${path.basename(outPath)}\n\n## Prompt used\n\n${prompt}\n`;
  fs.writeFileSync(sidecarPath, sidecar, 'utf8');

  console.log(
    `Generated ${imageBytes.length} bytes -> ${outPath} (sidecar: ${path.basename(sidecarPath)})`,
  );
}

main().catch((err) => {
  fail(err?.message ?? String(err));
});
